Added tutorial program step-82

dianeguignard opened the pull request:
In this program, we solve a bi-Laplacian problem using a _local discontinuous Galerkin_ (LDG) approach. Using so-called discrete/reconstructed Hessians, the discrete problem can be obtained by replacing the continuous Hessians appearing in the weak formulation by their discrete counter-part (plus penalty terms).

bangerth commented:
First, my apologies for letting this sit for so long :-( I've started to read through it. There isn't much that needs to be changed, but I'm collecting small things I'm finding while reading. Here is my first set of changes (to the introduction): https://github.com/bangerth/dealii/commit/3f49d6a3df4368046eaa7ca2f6b07db3910456cc
I tried to create a pull request against your own pull request, but github doesn't let me do this. We'll figure this out when I'm through.

In the meantime: Is there a publication that could be cited somewhere at the top that would provide some background knowledge? I suspect the two of you wrote something for which this program is simply the accompanying documentation :-)

bangerth commented:
Separately: In the results section, you only show results for k=2. For that case, the L2 error converges at sub-optimal rate 2 instead of 3. This actually nicely matches what we see in step-47. But in step-47, we demonstrate that for k=3, you recover the optimal rate in the L2 norm (namely, 4). Have you tried this for your program per chance?

dianeguignard commented:
> First, my apologies for letting this sit for so long :-( I've started to read through it. There isn't much that needs to be changed, but I'm collecting small things I'm finding while reading. Here is my first set of changes (to the introduction): [bangerth@3f49d6a](https://github.com/bangerth/dealii/commit/3f49d6a3df4368046eaa7ca2f6b07db3910456cc)
> I tried to create a pull request against your own pull request, but github doesn't let me do this. We'll figure this out when I'm through.
> 
> In the meantime: Is there a publication that could be cited somewhere at the top that would provide some background knowledge? I suspect the two of you wrote something for which this program is simply the accompanying documentation :-)

No problem and thanks a lot for your updates in the introduction. I may have one or two comments but I will wait until you are done with a first read to submit all of them at once.

Regarding publications, we have two submitted papers where we use the reconstructed Hessian to discretize the energy (bending energy for prestrained plate); preprints can be found on arXiv, see [2011.01086](https://arxiv.org/abs/2011.01086) and [2106.13877](https://arxiv.org/abs/2106.13877). We introduce the numerical method in the former, while the latter consists of the numerical analysis of the scheme. However, I am not sure that these articles would be useful to provide background on the subject as the problem is quite involved (but I think that they should be cited in the tutorial anyway, for instance when we mention the advantage of the LDG method, see the third item of the note in the section "LDG Approximations"). Before implementing the _prestrained plate problem_, we first considered the LDG method for the Poisson (reconstructed gradient) and bi-Laplacian (reconstructed Hessian) problems, and thought that it would be useful to make on tutorial on that. What do you think?

dianeguignard commented:
> Separately: In the results section, you only show results for k=2. For that case, the L2 error converges at sub-optimal rate 2 instead of 3. This actually nicely matches what we see in step-47. But in step-47, we demonstrate that for k=3, you recover the optimal rate in the L2 norm (namely, 4). Have you tried this for your program per chance?

I did some tests with different polynomial degrees, but I did not save the results. I just run the code with k=3 and we indeed get the expected convergence rates. The details are below.

`<table align="center" class="doxtable">
  <tr>
    <th>n_ref</th>
    <th>n_cells</th>
    <th>n_dofs</th>
    <th>error H2 </th>
    <th>rate</th>
    <th>error H1</th>
    <th>rate</th>
    <th>error L2</th>
    <th>rate</th>
  </tr>
  <tr>
    <td align="center">1</td>
    <td align="right">4</td>
    <td align="right">36</td>
    <td align="center">1.451e-02</td>
    <td align="center">--</td>
    <td align="center">5.494e-04</td>
    <td align="center">--</td>
    <td align="center">3.035e-05</td>
    <td align="center">--</td>    	
  </tr>
  <tr>
    <td align="center">2</td>
    <td align="right">16</td>
    <td align="right">144</td>
    <td align="center">3.565e-03</td>
    <td align="center">2.02</td>
    <td align="center">6.870e-05</td>
    <td align="center">3.00</td>
    <td align="center">2.091e-06</td>
    <td align="center">3.86</td>    
  </tr>
  <tr>
    <td align="center">3</td>
    <td align="right">64</td>
    <td align="right">576</td>
    <td align="center">8.891e-04</td>
    <td align="center">2.00</td>
    <td align="center">8.584e-06</td>
    <td align="center">3.00</td>
    <td align="center">1.352e-07</td>
    <td align="center">3.95</td>
  </tr>
  <tr>
    <td align="center">4</td>
    <td align="right">256</td>
    <td align="right">2304</td>
    <td align="center">2.223e-04</td>
    <td align="center">2.00</td>
    <td align="center">1.073e-06</td>
    <td align="center">3.00</td>
    <td align="center">8.594e-09</td>
    <td align="center">3.98</td>
  </tr>
  <tr>
    <td align="center">5</td>
    <td align="right">1024</td>
    <td align="right">9216</td>
    <td align="center">5.560e-05</td>
    <td align="center">2.00</td>
    <td align="center">1.341e-07</td>
    <td align="center">3.00</td>
    <td align="center">5.418e-10</td>
    <td align="center">3.99</td>
  </tr>
  <tr>
    <td align="center">6</td>
    <td align="right">4096</td>
    <td align="right">36864</td>
    <td align="center">1.390e-05</td>
    <td align="center">2.00</td>
    <td align="center">1.676e-08</td>
    <td align="center">3.00</td>
    <td align="center">3.245e-11</td>
    <td align="center">4.06</td>
  </tr>
</table>`

bangerth commented:
Ah, nice. Thanks for the experiments! I've put that table into my edits of the results section.

bangerth commented:
As for the references: Let's remember that for the end. I think it would be useful to add these references and be it only so that you can get the credit you deserve. But I'll let you add these references after pulling my changes onto your branch.

bangerth commented:
As for that last line: You have just reached round-off level.

dianeguignard commented:
> As for the references: Let's remember that for the end. I think it would be useful to add these references and be it only so that you can get the credit you deserve. But I'll let you add these references after pulling my changes onto your branch.

Sounds good.

bangerth commented:
I've made it through the rest of the program and only have minor changes to suggest, mostly to make the program look more like the others. 

Do you know how to pull the commits from my branch here: https://github.com/bangerth/dealii/commits/added-step-82 If not, let me know and I'll walk you through the process!

dianeguignard commented:
Thanks a lot Wolfgang for your edits. I read through them and made very few minor modifications in the introduction and the results sections. In particular:

- In the results, I removed the results for 7 refinements for the case `k=2` since we do not provide them for `k=3`.

- In the introduction, I added a reference to two preprints. I modified the "references.bib" file I had in my folder (fork of deal.ii on June 11) which is not in sych with the current version, but I guess this is not an issue for the merge. Also, is there a format I should use for citing arXiv preprints? There were two other occurrences (_clevenger_par_gmg_ and _munch2020hyperdeal_) which use different formats. In any case, the preprint number is not showing up in the bibliography (but this is a detail since the input will be updated once the articles are published).

bangerth commented:
Looks good, thanks!

For the references: If you add an `url = {...}` field to your entries, then there will at least be a link that people can follow.

Other than that, I have only one request: Could you please put an entry into the `doc/news/changes/major` directory in the same format as the other files in there? This way, your contribution shows up in the changelog file for the next release!

bangerth commented:
You will also want to address this compile failure:
```
[ 88%] Building CXX object examples/CMakeFiles/step-82.debug.dir/step-82/step-82.cc.o
/Users/runner/work/dealii/dealii/examples/step-82/step-82.cc:402:21: error: loop variable 'cell' of type 'const dealii::TriaActiveIterator<dealii::DoFCellAccessor<2, 2, false> >' creates a copy from type 'const dealii::TriaActiveIterator<dealii::DoFCellAccessor<2, 2, false> >' [-Werror,-Wrange-loop-analysis]
    for (const auto cell : dof_handler.active_cell_iterators())
                    ^
/Users/runner/work/dealii/dealii/examples/step-82/step-82.cc:1438:5: note: in instantiation of member function 'Step82::BiLaplacianLDGLift<2>::setup_system' requested here
    setup_system();
    ^
/Users/runner/work/dealii/dealii/examples/step-82/step-82.cc:1471:15: note: in instantiation of member function 'Step82::BiLaplacianLDGLift<2>::run' requested here
      problem.run();
              ^
/Users/runner/work/dealii/dealii/examples/step-82/step-82.cc:402:10: note: use reference type 'const dealii::TriaActiveIterator<dealii::DoFCellAccessor<2, 2, false> > &' to prevent copying
    for (const auto cell : dof_handler.active_cell_iterators())
         ^~~~~~~~~~~~~~~~~
                    &
1 error generated.
```
You should use `for (const auto &cell : ...)` -- note the `&`.

bangerth commented:
Awesome -- let's see what happens!

/rebuild

bangerth commented:
There is one failure that I believe is unrelated but that should be easy to check. Do you know how to rebase to master?

peterrum commented:
Let me also take a look (later today) at this tutorial before merging. Thanks!

peterrum commented on examples/step-82/step-82.cc:
```diff
@@ -0,0 +1,1500 @@
+      discrete_hessians_neigh(GeometryInfo<dim>::faces_per_cell,
+                              discrete_hessians);
+
+    Tensor<2, dim> H_i, H_j;
+    Tensor<2, dim> H_i_neigh, H_j_neigh;
+    Tensor<2, dim> H_i_neigh2, H_j_neigh2;
```
I don't think we need to declare these here.

dianeguignard replied:
Do you mean that we should directly use `Tensor <2, dim> H_i = discrete_hessians[i][q];` at line 556 and similarly for the other occurrences?

peterrum replied:
Yes. The current version is more `C`-style. Assigning a value at a single place allows you to make the variable const, which significantly improves the readability of codes since the origin of values is clear. I know that it is annoying to type out the type, but you could also use `const auto H_i = discrete_hessians[i][q];`

peterrum commented on examples/step-82/step-82.cc:
```diff
@@ -0,0 +1,1500 @@
+                  cell->neighbor_of_neighbor(face_no);
+
+                if (neighbor_cell->id().operator<(cell->id()))
+                  { // we need to have a global way to compare the cells in
+                    // order to not calculate the same jump term twice
+                    continue; // skip this face (already considered)
```
Doxygen cannot handle such comments.
```suggestion
                  { 
                    // we need to have a global way to compare the cells in
                    // order to not calculate the same jump term twice
                    // skip this face (already considered)
                    continue; 
```

dianeguignard replied:
Thanks for spotting this. What about using the same structure as in the function `compute_errors`, namely

```suggestion
                // In the next step, we need to have a global way to compare the
                // cells in order to not calculate the same jump term twice:
                if (neighbor_cell->id() < cell->id())
                    continue; // skip this face (already considered) 
```

peterrum replied:
Sounds good!

peterrum commented on examples/step-82/step-82.cc:
```diff
@@ -0,0 +1,1500 @@
+  template <int dim>
+  void BiLaplacianLDGLift<dim>::compute_errors()
+  {
+    double error_H2 = 0;
+    double error_H1 = 0;
+    double error_L2 = 0;
+
+    QGauss<dim>     quad(fe.degree + 1);
+    QGauss<dim - 1> quad_face(fe.degree + 1);
+
+    FEValues<dim> fe_values(fe,
```
I guess ones cannot use the `VectorTools` functions here because they do not consider faces(?). Would it be an option to create a new function in that namespace?

peterrum commented on examples/step-82/step-82.cc:
```diff
@@ -0,0 +1,1500 @@
+      coeffs_re(n_dofs_lift), coeffs_be(n_dofs_lift), coeffs_tmp(n_dofs_lift);
+
+    SolverControl solver_control(1000, 1e-12);
+    SolverCG<>    solver(solver_control);
```
May I ask you to make the following change:
```suggestion
    SolverCG<Vector<double>>   solver(solver_control);
```
We try to avoid `<>` in the tutorials.

dianeguignard replied:
Of course!

peterrum commented on examples/step-82/step-82.cc:
```diff
@@ -0,0 +1,1500 @@
+          discrete_hessians[i][q] = 0;
+
+          for (unsigned int face_no = 0;
+               face_no < GeometryInfo<dim>::faces_per_cell;
```
```suggestion
               face_no < discrete_hessians_neigh.size();
```

One `GeometryInfo` less :)

dianeguignard replied:
Thanks!

peterrum commented on examples/step-82/step-82.cc:
```diff
@@ -0,0 +1,1500 @@
+{
+  try
+    {
+      const unsigned int degree =
```
Would it be an option to make n_refinements also a parameter here. So one only has to modify the parameters at one place to reproduce the tables in the result section.

dianeguignard replied:
Sure, I made the update ;-).

peterrum commented on examples/step-82/step-82.cc:
```diff
@@ -0,0 +1,1500 @@
+                                              const double penalty_jump_val)
+    : fe(fe_degree)
+    , dof_handler(triangulation)
+    , fe_lift(FE_DGQ<dim>(fe_degree), dim * dim)
```
I think I missed the origin of `dim * dim`. Could you explain the origin. I guess it number of entries in the Hessian?

dianeguignard replied:
Indeed, it corresponds to the size of the Hessian. The discrete Hessian, denoted `H_h` in the introduction and defined such that it weakly converges to the continuous one `D^2`, consists of three parts: the broken Hessian plus the lifting of the jumps of the function and the jumps of the gradient of the function. All these terms are in `$\mathbb{R}^{d\times d}$`.

peterrum commented on examples/step-82/step-82.cc:
```diff
@@ -0,0 +1,1500 @@
+    : fe(fe_degree)
+    , dof_handler(triangulation)
+    , fe_lift(FE_DGQ<dim>(fe_degree), dim * dim)
+    , dof_handler_lift(triangulation)
```
If I understand it correctly, you technically don't need `dof_handler_lift`, but only use it initialize `FEValues` to be used locally. I am not sure but you could hand in to `FEValues::reinit()` a `CellAccessor` instead of a `DoFCellAccessor` and then it should work!? Maybe I missing something? Setting up a system that is 9 times bigger than the actual problem seems to be a pity.

bangerth replied:
And just to be explicit: when @peterrum says `CellAccessor`, what he means is `Triangulation<dim>::cell_iterator` (as opposed to `DoFHandler<dim>::cell_iterator`).

dianeguignard replied:
I am not sure I understood your comment correctly. We always have that `cell` and `cell_lift` refer to the same cell. However, the local problems we need to solve for the lifting terms (see `$M_cB_c=G_c$` in the introduction) indeed involve tensor-valued functions of size `dim * dim` (4 or 9). In other words, all the lifting terms belong to the FE space `$\mathbb{V}_h^{d\times d}$`.

For simplicity, in this program `$\mathbb{V}_h$` is the same FE space as the one we use for the solution (in particular we use the same FE degree), and we thus have `dim*dim` copies of the FE space we use for the solution. So I think that we could only use `fe` and `dof_handler` (i.e., remove `fe_lift` and `dof_handler_lift`), but then it would basically correspond to do by hand what is done by `FESystem`, which would complicate the description of the computation of `H_h` as well as the implementation...

Please let me know if I misunderstood something or if I am missing something.

peterrum replied:
I have opened a small PR that shows you how to remove the second DoFHandler: https://github.com/dianeguignard/dealii/pull/1

peterrum commented on examples/step-82/step-82.cc:
```diff
@@ -0,0 +1,1500 @@
+    typename DoFHandler<dim>::active_cell_iterator cell_lift =
+      dof_handler_lift.begin_active();
+
+    for (; cell != endc; ++cell, ++cell_lift)
```
If we can eliminate the second DoFHandler, I would suggest to loop over the active cells of the triangulation and create a `DoFCellAccessor` based on the `CellAcessor` for the scalar DoFHandler.

peterrum reviewed (commented):
I left some comments. Looks overall good (although I have not understood all details). My only concern is regarding the second DoFHandler, which might be eliminated if I am not mistaken.

dianeguignard commented on examples/step-82/doc/intro.dox:
```diff
@@ -0,0 +1,266 @@
+@f[
+{\rm compute\_discrete\_hessians[i][q]}, \qquad 0\leq {\rm i} < {\rm n\_dofs}, \,\, 0\leq {\rm q} < {\rm n\_q\_points},
+@f]
+where <code>n_dofs = fe_values.dofs_per_cell</code> is the number of degrees of freedom per cell and <code>n_q_points = quad.size()</code> is the number of quadrature points on $K_c$. For any basis function $\varphi^n$ with support on a neighboring cell, the discrete Hessian $H_h(\varphi^n)$ evaluated on $K_c$ contains only the two lifting terms, but not the term involving $D^2)h\varphi^n$, since ...
```
Indeed, thank you for spotting this. `h` should be a subscript, i.e., it should be `$D^2_h\varphi_n$` which is the broken Hessian of `\varphi_n`.

dianeguignard commented:
Do I need to do anything else at this point as some checks were not successful? In particular, there was a failure not related to this tutorial but to a Stokes problem (see Wolfgang's comment)...

masterleinad commented:
> Do I need to do anything else at this point as some checks were not successful? In particular, there was a failure not related to this tutorial but to a Stokes problem (see Wolfgang's comment)...

Can you run the indenting script `./contrib/utilities/indent`?

dianeguignard commented:
> > Do I need to do anything else at this point as some checks were not successful? In particular, there was a failure not related to this tutorial but to a Stokes problem (see Wolfgang's comment)...
> 
> Can you run the indenting script `./contrib/utilities/indent`?

I get the following error message 6 times when I run it (after running `./contrib/utilities/download_clang_format
`)

`fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
fatal: Not a valid object name `

masterleinad commented:
> I get the following error message 6 times when I run it (after running `./contrib/utilities/download_clang_format `)
> 
> `fatal: ambiguous argument 'master': unknown revision or path not in the working tree. Use '--' to separate paths from revisions, like this: 'git <command> [<revision>...] -- [<file>...]' fatal: Not a valid object name `

Hmm... It's difficult to diagnose remotely what the problem here is. I haven't seen that before.
Can you try `./contrib/utilities/indent-all` instead? This should be (much) slower but also help with the indentation here.

dianeguignard commented:
> > I get the following error message 6 times when I run it (after running `./contrib/utilities/download_clang_format `)
> > `fatal: ambiguous argument 'master': unknown revision or path not in the working tree. Use '--' to separate paths from revisions, like this: 'git <command> [<revision>...] -- [<file>...]' fatal: Not a valid object name `
> 
> Hmm... It's difficult to diagnose remotely what the problem here is. I haven't seen that before. Can you try `./contrib/utilities/indent-all` instead? This should be (much) slower but also help with the indentation here.

I could run this one and I have committed the updated version. It seems that the indentation is now fine.

peterrum reviewed (approved):
I took the liberty to rebase your branch. This failing test was fixed a long time ago on master. Hopefully, CI is now happy! If yes, this PR ready from my side! Thank you for the nice program; it was fun to read through it!

peterrum commented:
OK. CI is happy! Anyone else who is interested in reviewing this tutorial? @bangerth you have not approved yet!
